nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/iot/PutAWSIoT.java

```diff
@@ -0,0 +1,141 @@
+})
+@WritesAttributes({
+        @WritesAttribute(attribute = "aws.iot.mqtt.exception", description = "Error details")
+})
+public class PutAWSIoT extends AbstractAWSIoTProcessor {
+    private final static String PROP_NAME_RETAINED = "aws.iot.mqtt.retained";
+    private final static String ATTR_NAME_TOPIC = PROP_NAME_TOPIC + ".override";
+    private final static String ATTR_NAME_QOS = PROP_NAME_QOS + ".override";
+    private final static String ATTR_NAME_RETAINED = PROP_NAME_RETAINED + ".override";
+    private final static String ATTR_NAME_EXCEPTION = "aws.iot.mqtt.exception";
+    private final static Boolean PROP_DEFAULT_RETAINED = false;
+    private Boolean shouldRetain;
+
+    public static final PropertyDescriptor PROP_RETAINED = new PropertyDescriptor
+            .Builder().name(PROP_NAME_RETAINED)
+            .description("For messages being published, a true setting indicates that the MQTT server " +
+                    "should retain a copy of the message. The message will then be transmitted to new " +
+                    "subscribers to a topic that matches the message topic. For subscribers registering " +
+                    "a new subscription, the flag being true indicates that the received message is not " +
+                    "a new one, but one that has been retained by the MQTT server.")
+            .required(true)
+            .defaultValue(PROP_DEFAULT_RETAINED.toString())
+            .addValidator(StandardValidators.BOOLEAN_VALIDATOR)
+            .build();
+
+    public static final List<PropertyDescriptor> properties = Collections.unmodifiableList(
+            Arrays.asList(
+                    PROP_QOS,
+                    PROP_TOPIC,
+                    PROP_RETAINED,
+                    PROP_ENDPOINT,
+                    PROP_KEEPALIVE,
+                    PROP_CLIENT,
+                    AWS_CREDENTIALS_PROVIDER_SERVICE,
+                    TIMEOUT,
+                    REGION));
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return Collections.singleton(REL_SUCCESS);
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        shouldRetain = context.getProperty(PROP_RETAINED).isSet() ? context.getProperty(PROP_RETAINED).asBoolean() : PROP_DEFAULT_RETAINED;
+        init(context);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) {
+        // check if MQTT-connection is about to expire
+        if (isConnectionAboutToExpire()) {
+            // renew connection
+            mqttClient = connect(context);
+        }
+        // get flowfile
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
```

Comment:
This check for a flowfile should be before the reconnect. There is no point to reconnect if there is no FlowFile to process.